ledger-tool/src/blockstore.rs

```diff
@@ -0,0 +1,1127 @@
+            );
+        }
+        (slot, hash_and_timestamp_opt, contains_nonvote_tx)
+    });
+
+    if exclude_vote_only_slots {
+        slot_iter
+            .filter(|(_, _, contains_nonvote)| *contains_nonvote)
+            .take(num_slots)
+            .collect()
+    } else {
+        slot_iter.take(num_slots).collect()
+    }
+}
+
+fn print_blockstore_file_metadata(
+    blockstore: &Blockstore,
+    file_name: &Option<&str>,
+) -> Result<(), String> {
+    let live_files = blockstore
+        .live_files_metadata()
+        .map_err(|err| format!("{err:?}"))?;
+
+    // All files under live_files_metadata are prefixed with "/".
+    let sst_file_name = file_name.as_ref().map(|name| format!("/{name}"));
+    for file in live_files {
+        if sst_file_name.is_none() || file.name.eq(sst_file_name.as_ref().unwrap()) {
+            println!(
+                "[{}] cf_name: {}, level: {}, start_slot: {:?}, end_slot: {:?}, size: {}, \
+                 num_entries: {}",
+                file.name,
+                file.column_family_name,
+                file.level,
+                raw_key_to_slot(&file.start_key.unwrap(), &file.column_family_name),
+                raw_key_to_slot(&file.end_key.unwrap(), &file.column_family_name),
+                file.size,
+                file.num_entries,
+            );
+            if sst_file_name.is_some() {
+                return Ok(());
+            }
+        }
+    }
+    if sst_file_name.is_some() {
+        return Err(format!(
+            "Failed to find or load the metadata of the specified file {file_name:?}"
+        ));
+    }
+    Ok(())
+}
+
+pub trait BlockstoreSubCommand {
+    fn blockstore_subcommand(self) -> Self;
+}
+
+impl BlockstoreSubCommand for App<'_, '_> {
+    fn blockstore_subcommand(self) -> Self {
+        self.subcommand(
+            SubCommand::with_name("blockstore")
+                .about("Commands to interact with a local Blockstore")
+                .setting(AppSettings::InferSubcommands)
+                .setting(AppSettings::SubcommandRequiredElseHelp)
+                .subcommands(blockstore_subcommands(false)),
+        )
+    }
+}
+
+pub fn blockstore_subcommands<'a, 'b>(hidden: bool) -> Vec<App<'a, 'b>> {
+    let hidden = if hidden {
+        vec![AppSettings::Hidden]
+    } else {
+        vec![]
+    };
+
+    let starting_slot_arg = Arg::with_name("starting_slot")
+        .long("starting-slot")
+        .value_name("SLOT")
+        .takes_value(true)
+        .default_value("0")
+        .help("Start at this slot");
+    let ending_slot_arg = Arg::with_name("ending_slot")
+        .long("ending-slot")
+        .value_name("SLOT")
+        .takes_value(true)
+        .help("The last slot to iterate to");
+    let allow_dead_slots_arg = Arg::with_name("allow_dead_slots")
+        .long("allow-dead-slots")
+        .takes_value(false)
+        .help("Output dead slots as well");
```

Comment:
Could we/should we dedupe these with the existing definitions in main.rs at some point?

Reply:
Yeah definitely; it is not expressed here, but I loosely allude to wanting to do some reorganization on that front in the problem descriptions for https://github.com/solana-labs/solana/pull/34694 (a PR I have yet to add you too 😉).

The goals there would be deduping argument definitions (there were a few dupes prior to my changes here), and also creating several groupings of arguments. The groupings could be added to subcommands via [this function](https://docs.rs/clap/2.34.0/clap/struct.App.html#method.args) to avoid long repeated lists like this for the `[verify, create-snapshot, graph, ...]` commands:
https://github.com/solana-labs/solana/blob/843854409ee9c13fe548f44e2ad2d555703cfb17/ledger-tool/src/main.rs#L1368-L1392